File: backend/services/soilgrids_service.py
import json
import urllib.request
import urllib.parse
import asyncio
from typing import Dict, Any
from backend.config import settings

class SoilGridsService:
    def __init__(self, base_url: str = settings.SOILGRIDS_API_URL):
        self.base_url = base_url

    def _sync_fetch(self, url: str) -> Dict[str, Any]:
        req = urllib.request.Request(url, headers={"User-Agent": "AgroXAI-Service/1.0", "Accept": "application/json"})
        with urllib.request.urlopen(req, timeout=6.0) as resp:
            return json.loads(resp.read().decode("utf-8"))

    async def get_soil_properties(self, lat: float, lon: float) -> Dict[str, Any]:
        """
        Query SoilGrids 2.0 REST API for pH, organic carbon, and clay percentage at 0-30cm depth.
        """
        params = [
            ("lon", str(lon)),
            ("lat", str(lat)),
            ("property", "phh2o"),
            ("property", "soc"),
            ("property", "clay"),
            ("depth", "0-30cm"),
            ("value", "mean"),
        ]
        query_string = urllib.parse.urlencode(params)
        url = f"{self.base_url}?{query_string}"
        
        try:
            data = await asyncio.to_thread(self._sync_fetch, url)
            layers = data.get("properties", {}).get("layers", [])
            
            # Extract phh2o
            ph_layer = next((l for l in layers if l.get("name") == "phh2o"), None)
            raw_ph = ph_layer.get("depths", [{}])[0].get("values", {}).get("mean") if ph_layer else None
            ph = round(raw_ph / 10.0, 1) if raw_ph else 7.2
            
            # Extract soc
            soc_layer = next((l for l in layers if l.get("name") == "soc"), None)
            raw_soc = soc_layer.get("depths", [{}])[0].get("values", {}).get("mean") if soc_layer else None
            soc = round(raw_soc / 100.0, 1) if raw_soc else 0.8
            
            # Extract clay
            clay_layer = next((l for l in layers if l.get("name") == "clay"), None)
            raw_clay = clay_layer.get("depths", [{}])[0].get("values", {}).get("mean") if clay_layer else None
            clay = round(raw_clay / 10.0, 1) if raw_clay else 48.0
            
            soil_type = "Deep Black Soil (Vertisol)" if clay > 40 else "Alluvial Loam (Inceptisol)"
            
            return {
                "success": True,
                "source": "soilgrids-rest",
                "soilType": soil_type,
                "ph": ph,
                "organicCarbonPercent": soc,
                "clayContentPercent": clay,
                "nitrogenBaseline": 70 if ph > 7.5 else 85,
                "phosphorusBaseline": 42,
                "potassiumBaseline": 48 if ph > 7 else 38,
            }
        except Exception as e:
            # Regional agronomic fallback
            return {
                "success": True,
                "source": "regional-classification",
                "soilType": "Deep Black Soil (Vertisol)",
                "ph": 7.5,
                "organicCarbonPercent": 0.75,
                "clayContentPercent": 52.0,
                "nitrogenBaseline": 75,
                "phosphorusBaseline": 44,
                "potassiumBaseline": 50,
                "error": str(e),
            }

soilgrids_service = SoilGridsService()

File: src/services/soilGridsService.ts
/**
 * ISRIC SoilGrids 2.0 Integration Service
 * Queries global digital soil mapping for pH, organic carbon, and texture classifications.
 */

export interface SoilGridsData {
  success: boolean;
  source: 'soilgrids-rest' | 'regional-classification';
  soilType: string;
  ph: number;
  nitrogenBaseline: number; // estimated mg/kg
  phosphorusBaseline: number;
  potassiumBaseline: number;
  organicCarbonPercent: number;
  clayContentPercent: number;
  description: string;
}

export class SoilGridsService {
  private baseUrl: string;

  constructor(
    baseUrl: string = 'https://rest.isric.org/soilgrids/v2.0/properties/query'
  ) {
    this.baseUrl = baseUrl;
  }

  /**
   * Query SoilGrids REST API for baseline chemistry at depth 0-30cm.
   */
  async getSoilProperties(lat: number, lon: number): Promise<SoilGridsData> {
    const url = `${this.baseUrl}?lon=${lon}&lat=${lat}&property=phh2o&property=soc&property=clay&depth=0-30cm&value=mean`;

    try {
      const response = await fetch(url, {
        headers: { Accept: 'application/json' },
      });

      if (!response.ok) {
        throw new Error(`SoilGrids API returned status ${response.status}`);
      }

      const data = await response.json();
      const layers = data.properties?.layers || [];

      // Extract phh2o (SoilGrids returns pH * 10)
      const phLayer = layers.find((l: any) => l.name === 'phh2o');
      const rawPh = phLayer?.depths?.[0]?.values?.mean;
      const ph = rawPh ? Math.round((rawPh / 10) * 10) / 10 : 7.2;

      // Extract soc (soil organic carbon dg/kg)
      const socLayer = layers.find((l: any) => l.name === 'soc');
      const rawSoc = socLayer?.depths?.[0]?.values?.mean;
      const socPercent = rawSoc ? Math.round((rawSoc / 100) * 10) / 10 : 0.8;

      // Extract clay (% clay particles)
      const clayLayer = layers.find((l: any) => l.name === 'clay');
      const rawClay = clayLayer?.depths?.[0]?.values?.mean;
      const clayPercent = rawClay ? Math.round((rawClay / 10) * 10) / 10 : 48;

      // Infer regional soil type based on coordinates and clay content
      const soilType = clayPercent > 40 ? 'Black Cotton Soil (Vertisol)' : 'Alluvial Loam (Inceptisol)';

      return {
        success: true,
        source: 'soilgrids-rest',
        soilType,
        ph,
        nitrogenBaseline: ph > 7.5 ? 65 : 85,
        phosphorusBaseline: 42,
        potassiumBaseline: ph > 7 ? 48 : 38,
        organicCarbonPercent: socPercent,
        clayContentPercent: clayPercent,
        description: `High-retention ${soilType.toLowerCase()} with deep profile and moderate alkaline base.`,
      };
    } catch (error) {
      console.warn('SoilGridsService using regional agronomic classification fallback:', error);
      return this.getRegionalClassification(lat, lon);
    }
  }

  /**
   * Deterministic regional agronomic soil classification fallback based on Indian subcontinental / global coordinates.
   */
  getRegionalClassification(lat: number, lon: number): SoilGridsData {
    // Deccan Trap Vertisol zone check
    const isDeccanPlateau = lat >= 15.0 && lat <= 21.0 && lon >= 73.0 && lon <= 78.0;

    if (isDeccanPlateau) {
      return {
        success: true,
        source: 'regional-classification',
        soilType: 'Deep Black Soil (Vertisol)',
        ph: 7.6,
        nitrogenBaseline: 70,
        phosphorusBaseline: 45,
        potassiumBaseline: 52,
        organicCarbonPercent: 0.72,
        clayContentPercent: 54,
        description: 'Montmorillonite-rich vertisol with high moisture retention and self-mulching structure.',
      };
    }

    return {
      success: true,
      source: 'regional-classification',
      soilType: 'Alluvial Loam (Inceptisol)',
      ph: 6.8,
      nitrogenBaseline: 90,
      phosphorusBaseline: 48,
      potassiumBaseline: 42,
      organicCarbonPercent: 0.85,
      clayContentPercent: 28,
      description: 'Well-drained alluvial fertile loam with balanced nutrient reserves.',
    };
  }
}

export const soilGridsService = new SoilGridsService();
